### ISSUE TITLE
`kubeadm` package should depend on the same version of `kubelet` and `kubectl` packages

### ISSUE THREAD

**Issue by paralika** [kubernetes/kubernetes#72871]
**What happened**:
`kubeadm` package always install latest version of `Kubelet` and `kubectl` packages. 

```
root@node01:/var/cache/apt/archives# apt install kubeadm=1.11.6-00 --download-only
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following additional packages will be installed:
  cri-tools ebtables kubectl kubelet kubernetes-cni socat
The following NEW packages will be installed:
  cri-tools ebtables kubeadm kubectl kubelet kubernetes-cni socat
0 upgraded, 7 newly installed, 0 to remove and 11 not upgraded.
Need to get 48.0 MB of archives.
After this operation, 285 MB of additional disk space will be used.
Do you want to continue? [Y/n] Y
Get:1 http://asia-south1-a.gce.clouds.archive.ubuntu.com/ubuntu xenial-updates/main amd64 ebtables amd64 2.0.10.4-3.4ubuntu2.16.04.2 [79.9 kB]
Get:2 https://apt.kubernetes.io kubernetes-xenial/main amd64 cri-tools amd64 1.12.0-00 [5,343 kB]
Get:3 http://asia-south1-a.gce.clouds.archive.ubuntu.com/ubuntu xenial/universe amd64 socat amd64 1.7.3.1-1 [321 kB]
Get:4 https://apt.kubernetes.io kubernetes-xenial/main amd64 kubernetes-cni amd64 0.6.0-00 [5,910 kB]
Get:5 https://apt.kubernetes.io kubernetes-xenial/main amd64 kubelet amd64 1.13.2-00 [19.0 MB]
Get:6 https://apt.kubernetes.io kubernetes-xenial/main amd64 kubectl amd64 1.13.2-00 [7,847 kB]
Get:7 https://apt.kubernetes.io kubernetes-xenial/main amd64 kubeadm amd64 1.11.6-00 [9,427 kB]
Fetched 48.0 MB in 6s (7,576 kB/s)                                                                                                                                                     
Download complete and in download only mode
```
**What you expected to happen**:
`kubeadm` package should be depend on the same version of `kubelet` and `kubectl`. 
Above command should download `kubeadm 1.11.6`, `Kubelets 1.11.6` and `kubectl 1.11.6`. 

**How to reproduce it (as minimally and precisely as possible)**:
Add Kubernetes repository (`deb https://apt.kubernetes.io/ kubernetes-xenial main`) in source.list.d.
`apt-get update`
`apt install kubeadm=1.11.6-00 --download-only`

**Anything else we need to know?**:
I believe we need to update the dependancy of `kubeadm` package. 

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration: GCP VM instance
- OS (e.g. from /etc/os-release): `Ubuntu 16.04.5 LTS`
- Kernel (e.g. `uname -a`): `Linux node01 4.15.0-1026-gcp #27~16.04.1-Ubuntu SMP Fri Dec 7 09:59:47 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux`
- Install tools:
- Others:

**Comment by neolit123** [comment]
@kubernetes/sig-cluster-lifecycle 
cc @mauilion @chuckha

**Comment by dixudx** [comment]
Two  options for this:

1. install `kubelet` and `kubectl` with targeted version first, and then `kubeadm`;
2. bumped `kubelet` and `kubectl` with exact the same version with `kubeadm`;

Option 1 is doable;
Option 2 is not preferable. From the view of packaging, I don't think it's a good idea to have such a strict binding.

**Comment by paralika** [comment]
Not strict binding but if we could add some range or `OR` operator would be helpful. 
Something like:  `1.11.6  || >= 1.11.6` for `kubeadm 1.11.6`.

**Comment by rosti** [comment]
I am in favor of strict binding, since mismatching (even at patch level) components can cause weird errors. And checking each package version is not always on users minds.

**Comment by hoegaarden** [comment]
- Would it help (and is it even possible?) to set `Recommends` to `1.11.6` in addition to the already existing `>= 1.116`?
  - I can try to experiment with that
- How are the rpms setting up the dependencies? If the above helps, can we replicate that to the rpm world?
  - I currently have no context on rpms at all ...

https://github.com/kubernetes/kubernetes/issues/71677 might also be interesting in this context as it might change how packages are built.

@mariantalla

**Comment by paralika** [comment]
Yes. I think for `kubeadm` it should install dependancy with same version. If it is not found it may search for one level up and so on.

**Comment by liggitt** [comment]
is there a way to express the range of kubelet versions that are acceptable, which one to use if none is already present (probably latest matching patch release?), and to no-op if an acceptable version is already present?

**Comment by hoegaarden** [comment]
> is there a way to express the range of kubelet versions that are acceptable, which one to use if none is already present (probably latest matching patch release?), and to no-op if an acceptable version is already present?

That's what I try to do with setting `Depends` ("... an absolute dependency") to a range and `Recommends` ("... a strong, but not absolute, dependency") to a specific version, see in [the debian docs](https://www.debian.org/doc/debian-policy/ch-relationships.html#binary-dependencies-depends-recommends-suggests-enhances-pre-depends). However I have no idea how a mix of both `Depends` and `Recommends` really turns out and thus would like to test that.

**Comment by hoegaarden** [comment]
> is there a way to express the range of kubelet versions that are acceptable, which one to use if none is already present (probably latest matching patch release?), and to no-op if an acceptable version is already present?

I did some more tests. I think it is not possible to do something like that only with setting `Depends` and other fields on the packages. It might be possible to do something like this via setting up [pinning](https://manpages.debian.org/stretch/apt/apt_preferences.5.en.html#The_Effect_of_APT_Preferences). However packages [*should not override local administrator choices*](https://lintian.debian.org/tags/package-installs-apt-preferences.html) -- even more so when e.g. the `kubelet` package sets up pinning for the `kubeadm` package, as I tested [over here](https://github.com/pivotal-k8s/deb-ver-dep-experiments).

**Comment by justaugustus** [comment]
cc: @kubernetes/release-engineering 
Please investigate if this is still valid.

**Comment by neolit123** [comment]
/remove-priority important-soon
moving this to k/release
